src/metabase/query_processor/parameters.clj

```diff
@@ -0,0 +1,228 @@
+        formatter (tf/formatter "YYYY-MM-dd" tz)
+        today     (t/today-at-midnight tz)]
+    (->> (condp = value
+           "past7days"  {:end   (t/minus today (t/days 1))
+                         :start (t/minus today (t/days 7))}
+           "past30days" {:end   (t/minus today (t/days 1))
+                         :start (t/minus today (t/days 30))}
+           "thisweek"   (week-range today)
+           "thismonth"  (month-range today)
+           "thisyear"   (year-range today)
+           "lastweek"   (week-range (t/minus today (t/weeks 1)))
+           "lastmonth"  (month-range (t/minus today (t/months 1)))
+           "lastyear"   (year-range (t/minus today (t/years 1)))
+           "yesterday"  {:end   (t/minus today (t/days 1))
+                         :start (t/minus today (t/days 1))}
+           "today"      {:end   today
+                         :start today})
+         ;; the above values are JodaTime objects, so unparse them to iso8601 strings
+         (m/map-vals (partial tf/unparse formatter)))))
+
+(defn- extract-dates [value report-timezone]
+  (if-not (contains? relative-dates value)
+    ;; absolute date range
+    (absolute-date->range value)
+    ;; relative date range
+    (relative-date->range value report-timezone)))
+
+(defn- expand-date-range-param [report-timezone {[target param-name] :target, param-type :type, param-value :value, :as param}]
+  (if-not (= param-type "date")
+    param
+    (let [{:keys [start end]} (extract-dates param-value report-timezone)]
+      [(assoc param :target [target (str param-name ":start")], :value start)
+       (assoc param :target [target (str param-name ":end")],   :value end)])))
+
+(defn- substitute-param [param-name value query]
+  ;; TODO: escaping and protection against SQL injection!
+  (s/replace query (re-pattern (str "\\{\\{" param-name "\\}\\}")) value))
+
+(defn- substitute-all-params [query-dict [{:keys [value], [_ param-name] :target, :as param} & rest]]
+  (if param
+    (if-not (and param param-name value (string? param-name))
+      (substitute-all-params query-dict rest)
+      (let [query (update-in query-dict [:native :query] (partial substitute-param param-name value))]
+        (substitute-all-params query rest)))
+    query-dict))
+
+(def ^:private ^:const outer-clause #"\[\[.*?\]\]")
+(def ^:private ^:const outer-clause-prefix #"^\[\[(.*?)\s.*\]\]$")
+(def ^:private ^:const incomplete-outer-clause #"\[\[.*?\{\{.*?\}\}.*?\]\]")
+(def ^:private ^:const inner-clause #"<(.*?)>")
+(def ^:private ^:const incomplete-inner-clause #"<.*?\{\{.*?\}\}.*?>")
+
+(defn- remove-incomplete-clauses [query-dict]
+  (let [find-and-replace (fn [sql]
+                           (-> sql
+                               (s/replace incomplete-outer-clause "")
+                               (s/replace incomplete-inner-clause "")))]
+    (update-in query-dict [:native :query] find-and-replace)))
+
+(defn- conjoin-multi-clause [clause]
```

Comment:
It never hurts to write a quick blurb about the purpose of these functions. They might make sense to you, but I have no idea what they're for or what they're supposed to do.